---
title: "Breathing Together - Group 1"
author: "Céline Pattaroni"
date: "`r format(Sys.Date())`"
output:
  html_document:
    theme: journal
    fig_width: 10
    fig_height: 10
  pdf_document: default
editor_options:
  markdown:
    wrap: 100
---

Copyright (c) 2024, Computational Immunology Lab, Monash University, Melbourne, Australia.

### Environment set up

```{r, eval=FALSE}
# List of required packages
packages <- c(
  "biomaRt", "clusterProfiler", "decontam", "dplyr", "edgeR", "ggcorrplot", 
  "ggplot2", "ggpubr", "ggrepel", "ggside", "grDevices", "hydroTSM", "limma", "lubridate", 
  "magrittr", "matrixTests", "metagenomeSeq", "microbiome", "MOFA2", 
  "MultiAssayExperiment", "paletteer", "phyloseq", "reshape2", "Seurat", 
  "sctransform", "survival", "survminer", "tidyverse", "vegan", "xlsx"
)

# Install CRAN packages
install.packages(setdiff(packages, c("metagenomeSeq", "phyloseq", "clusterProfiler", "microbiome", "paletteer")))

# Install Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("metagenomeSeq", "phyloseq", "clusterProfiler"))

# Install GitHub packages
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("microbiome/microbiome")
#devtools::install_github("EmilHvitfeldt/paletteer")

# Add memory
usethis::edit_r_environ() #R_MAX_VSIZE=14GB

```

```{r}
# Set the checkpoint date for reproducibility
# Uncomment the following lines once the pipeline is ready
# library(checkpoint)
# checkpoint("2024-05-28")

# Load required libraries
library(microeco)
library(biomaRt)
library(clusterProfiler)
library(decontam)
library(dplyr)
library(edgeR)
library(ggcorrplot)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(ggside)
library(grDevices)
library(hydroTSM)
library(limma)
library(lubridate)
library(magrittr)
library(matrixTests)
library(metagenomeSeq)
library(microbiome)
library(MOFA2)
library(MultiAssayExperiment)
library(reshape2)
library(Seurat)
library(sctransform)
library(survival)
library(survminer)
library(tidyverse)
library(vegan)
library(xlsx)
library(file2meco)
library(dplyr)
library(gtsummary)
library(writexl)
library(segmented)
library(paletteer)
library(circlize)
library(ComplexHeatmap)
library(phyloseq)

# Remove phyloseq warnings
options(getClass.msg = FALSE)

# Set up global plot style using ggplot2
theme_set(theme_pubr(base_size = 10, base_family = "Helvetica") + 
          theme(plot.margin = unit(c(1, 1, 1, 1), "cm")))

# Set random seed for reproducibility
set.seed(2)

# Import custom functions from Functions.R
source("Functions.R")

# Set custom color palettes
custom_palette_2groups <- c("cadetblue4", "lightsalmon2")
custom_palette_age <- c("pink", "plum3", "plum4", "#66475A")
custom_palette_microbes <- c("gold2", "#D69F3DFF", "#825716FF")
custom_palette_omics1 <- c("pink3", "gold2", "#D69F3DFF", "#825716FF")
custom_palette_omics2 <- c("pink3", "#C83D43FF", "gold2", "#D69F3DFF", "#825716FF")
```

# [Metadata]{.underline}

#### 1) Metadata import

Metadata is imported and grouping is based on parents reported wheeze at year 1.

```{r}
# Import metadata
#source("Archive/Metadata.R") # REMOVE
#saveRDS(metadata, "Input/metadata_final.rds") # REMOVE
metadata <- readRDS("Input/metadata_final.rds")

# Set up group
metadata$Group <- metadata$Wheeze1YEAR
```

#### 2) Covariates

The analysis selects and tests a variety of covariates—ranging from environmental and clinical to
genetic and birth-related factors—across different groups for Supplementary Table 1. Chi-squared and
Wilcoxon tests are used to compare the different groups.

### ***Supplementary table 2***

```{r}
# Create lists of covariates (categorical)
covariates.env <- c("DayCare1YEAR", "NbSiblingsSimplified", "Pets", "Season1Y", "Country")
covariates.clinical <- c("Colds1YEARSimplified", "AbxSinceBorn1YEARSimplified")
covariates.genetic <- c("Gender", "MotherHayfever", "FatherHayfever", "MotherAsthma", "FatherAsthma")
covariates.birth <- c("DeliverySimplified", "GestAgeBirthSimplified", "WeightBirthKgSimplified", "AbxCoursesPregnancySimplified")
covariates.all <- c(covariates.env, covariates.clinical, covariates.genetic, covariates.birth)

# Filter metadata for relevant columns using any_of() to avoid errors
m <- metadata %>%
  dplyr::select(Wheeze1YEAR, any_of(c(covariates.all))) %>%
  filter(!is.na(Wheeze1YEAR) & Wheeze1YEAR != "Unknown")

# Ensure only existing columns are selected
existing_covariates <- covariates.all[covariates.all %in% colnames(m)]

# Create a summary table using gtsummary
summary_table <- m %>%
  dplyr::select(Wheeze1YEAR, any_of(existing_covariates)) %>%
  tbl_summary(
    by = Wheeze1YEAR,
    type = list(all_categorical() ~ "categorical"),
    statistic = list(all_categorical() ~ "{n} / {N} ({p}%)"),
    missing = "ifany"  # Include missing values in the summary
  ) %>%
  add_p(
    test = list(
      all_categorical() ~ "chisq.test"
    ),
    pvalue_fun = function(x) style_pvalue(x, digits = 3)
  ) %>%
  modify_header(label ~ "**Variable**") %>%
  bold_labels()

# Convert the gtsummary table to a data frame
summary_df <- as.data.frame(summary_table)

# Create a new Excel workbook and add the summary table
write_xlsx(summary_df, "Output/Supplementary_table_2.xlsx")
```

# [Microbiome data import]{.underline}

#### 1) Microbial data import

```{r}
# Load count tables, taxonomy, and trees for bacteria
bact.seqtab <- t(data.frame(readRDS("Input/16S_seqtab_nochim.rds"))) # Load count table
colnames(bact.seqtab) <- gsub("16S.", "", colnames(bact.seqtab)) # Rename columns
bact.taxonomy <- data.frame(readRDS("Input/16S_taxonomy_species.rds")) # Load taxonomy table
bact.taxonomy$ASV <- highest_taxClass(bact.taxonomy) # Add a column with the highest taxonomic class for each ASV
bact.tree <- readRDS("Input/16S_tree.rds") # Load phylogenetic tree

# Load count tables, taxonomy, and trees for fungi
fung.seqtab <- t(data.frame(readRDS("Input/ITS_seqtab_nochim.rds"))) # Load count table
colnames(fung.seqtab) <- gsub("ITS.", "", colnames(fung.seqtab)) # Rename columns
fung.taxonomy <- data.frame(readRDS("Input/ITS_taxonomy.rds")) # Load taxonomy table
fung.taxonomy$ASV <- highest_taxClass(fung.taxonomy) # Add a column with the highest taxonomic class for each ASV
fung.seq <- rownames(fung.taxonomy) # Save sequence IDs
fung.tree <- readRDS("Input/ITS_tree.rds") # Load phylogenetic tree

# Reshape taxonomy comparable to bacteria
fung.taxonomy <- data.frame(lapply(fung.taxonomy, function(x) gsub("k__|p__|c__|o__|f__|g__|s__", "", x))) # Remove prefix from taxonomy labels
rownames(fung.taxonomy) <- fung.seq # Rename rows with sequence IDs

# Load viral data
virus.tab <- read.xlsx2("Input/Viral_qPCR_results_summary.xlsx", colIndex=c(1:3), sheetIndex=3) # Load viral data from Excel file
virus.tab <- reshape::cast(virus.tab, Target~Sample, value="CT") # Reshape data to wide format
rownames(virus.tab) <- virus.tab$Target # Rename rows with viral targets
virus.tab <- virus.tab[,-1] # Remove the first column (redundant with row names)
virus.tab <- ifelse(virus.tab=="Undetermined" | virus.tab>40, 0, 1) # Replace non-detects with zeros and detectable values with ones

# Matching metadata
source("Archive/Metadata_microbiome.R") # REMOVE
microbiome.metadata <- df # REMOVE
saveRDS(microbiome.metadata, file = "Input/microbiome.metadata.rds") # REMOVE
df <- readRDS("Input/microbiome.metadata.rds")
```

#### 2) Phyloseq objects

```{r}
# Match metadata order with bacterial data
bact.seqtab <- bact.seqtab[, colnames(bact.seqtab) %in% rownames(df)]
df.bact <- df[rownames(df) %in% colnames(bact.seqtab), ]
df.bact <- df.bact[order(match(rownames(df.bact), colnames(bact.seqtab))), ]

# Match metadata order with fungal data
fung.seqtab <- fung.seqtab[, colnames(fung.seqtab) %in% rownames(df)]
df.fung <- df[rownames(df) %in% colnames(fung.seqtab), ]
df.fung <- df.fung[order(match(rownames(df.fung), colnames(fung.seqtab))), ]

# Match metadata order with viral data
df.virus <- df[df$VirusBC %in% colnames(virus.tab), ]
rownames(df.virus) <- df.virus$VirusBC
df.virus <- df.virus[order(match(rownames(df.virus), colnames(virus.tab))), ]

virus.tax <- as.data.frame(cbind(Kingdom = rep("Virus", nrow(virus.tab)), Species = rownames(virus.tab)))
rownames(virus.tax) <- virus.tax$Species

# Create phyloseq objects
bact <- phyloseq(
  otu_table(bact.seqtab, taxa_are_rows = TRUE),
  sample_data(df.bact),
  tax_table(as.matrix(bact.taxonomy)),
  phy_tree(bact.tree$tree)
)

fung <- phyloseq(
  otu_table(fung.seqtab, taxa_are_rows = TRUE),
  sample_data(df.fung),
  tax_table(as.matrix(fung.taxonomy)),
  phy_tree(fung.tree$tree)
)

vir <- phyloseq(
  otu_table(virus.tab, taxa_are_rows = TRUE),
  sample_data(df.virus),
  tax_table(as.matrix(virus.tax))
)

# Replace ASVs by taxa names (ID column)
taxa_names(bact) <- tax_table(bact)[, "ASV"]
taxa_names(fung) <- tax_table(fung)[, "ASV"]
```

# [Microbiome data pre-processing]{.underline}

#### 1) Samples filtering based on minimal counts

```{r}
minreadsTreshold <- 500

# Remove very low counts samples in bact
nsamples(bact)
bact <- prune_samples(colSums(otu_table(bact)) > minreadsTreshold, bact)
nsamples(bact)

# Remove very low counts samples in fung
nsamples(fung)
fung <- prune_samples(colSums(otu_table(fung)) > minreadsTreshold, fung)
nsamples(fung)

# Number of negative controls before
table(df$NegCtrl)

# Number of negative controls after
table(sample_data(bact)$NegCtrl)
table(sample_data(fung)$NegCtrl)
```

#### 2) Unclassified ASVs removal

```{r}
# Remove unclassifies ASVs in bact
ntaxa(bact)
bact <- subset_taxa(bact, !is.na(Phylum))
ntaxa(bact)

# Remove unclassifies ASVs in fung
ntaxa(fung)
fung <- subset_taxa(fung, !is.na(Phylum))
ntaxa(fung)
```

#### 3) ASVs decontamination

```{r}
# Decontamination of bact
bact.contamdf.prev <- isContaminant(bact, method = "prevalence", neg = "NegCtrl", threshold = 0.3)
bact.thresh <- ggplot(bact.contamdf.prev, aes(x = p)) +
  geom_density(fill = adjustcolor(c("darkseagreen"), 0.5), color = "darkseagreen", adjust = 5) +
  geom_vline(xintercept = 0.3, color = "indianred4") +
  ggtitle("Bacteria p density plot")
bact.contam <- bact.contamdf.prev[which(bact.contamdf.prev$contaminant == TRUE),]

# Decontamination of fung
fung.contamdf.prev <- isContaminant(fung, method = "prevalence", neg = "NegCtrl", threshold = 0.3)
fung.thresh <- ggplot(fung.contamdf.prev, aes(x = p)) +
  geom_density(fill = adjustcolor(c("darkseagreen"), 0.5), color = "darkseagreen", adjust = 5) +
  geom_vline(xintercept = 0.3, color = "indianred4") +
  ggtitle("Fungi p density plot")
fung.contam <- fung.contamdf.prev[which(fung.contamdf.prev$contaminant == TRUE),]

# Make a new phyloseq object of presence-absence in negative controls and true samples in bact
bact_pa <- transform_sample_counts(bact, function(abund) 1 * (abund > 0))
bact_pa_neg <- prune_samples(sample_data(bact_pa)$NegCtrl == TRUE, bact_pa)
bact_pa_pos <- prune_samples(sample_data(bact_pa)$NegCtrl == FALSE, bact_pa)

# Make a new phyloseq object of presence-absence in negative controls and true samples in fung
fung_pa <- transform_sample_counts(fung, function(abund) 1 * (abund > 0))
fung_pa_neg <- prune_samples(sample_data(fung_pa)$NegCtrl == TRUE, fung_pa)
fung_pa_pos <- prune_samples(sample_data(fung_pa)$NegCtrl == FALSE, fung_pa)

# Plot contaminants prevalence
contam.plot.bact <- PlotContaminants(bact, bact.contamdf.prev, "Bacteria contaminants ASVs")
contam.plot.fung <- PlotContaminants(fung, fung.contamdf.prev, "Fungi contaminants ASVs")

# Remove contaminant taxa in bact
ntaxa(bact)
bact <- prune_taxa(!(taxa_names(bact) %in% rownames(bact.contam)), bact)
ntaxa(bact)

# Remove contaminant taxa in fung
ntaxa(fung)
fung <- prune_taxa(!(taxa_names(fung) %in% rownames(fung.contam)), fung)
ntaxa(fung)

# Remove control samples
bact <- prune_samples(as.vector(!(sample_data(bact)[, "NegCtrl"])), bact)
fung <- prune_samples(as.vector(!(sample_data(fung)[, "NegCtrl"])), fung)

# Create figure
ggarrange(bact.thresh, contam.plot.bact, fung.thresh, contam.plot.fung,
  ncol = 2, nrow = 2, widths = c(1.5, 2)) + theme(text = element_text(size = 2))
```

#### 4) ASVs agglomeration (tree-based)

```{r eval=FALSE}
# Keep track of objects before agglomeration
bact_noagglo <- bact
fung_noagglo <- fung

# Agglomerate ASVs
bact <- tip_glom(bact, h = 0.05)
STOP
fung <- tip_glom(fung, h = 0.05)

# Add full taxonomic information again after tip_glom
bact <- addtaxonomyafteraggregation(bact)
fung <- addtaxonomyafteraggregation(fung)

# Save objects
saveRDS(bact, "Input/bact.rds")
saveRDS(fung, "Input/fung.rds")
```

#### 5) Prevalence filtering

```{r}
# Read rds
bact <- readRDS("Input/bact.rds")
fung <- readRDS("Input/fung.rds")
  
# Set prevalence filter parameters
prevalenceTheshold <- 0.05
mincountsThreshold <- 1

# Prevalence filtering in bact
ntaxa(bact)
bact <- filter_taxa(bact, function(x) sum(x>mincountsThreshold) > (prevalenceTheshold*length(x)), TRUE)
ntaxa(bact)

# Prevalence filtering in fung
ntaxa(fung)
fung <- filter_taxa(fung, function(x) sum(x>mincountsThreshold) > (prevalenceTheshold*length(x)), TRUE)
ntaxa(fung)
```

#### 6) Samples filtering based on read counts

```{r}
# Set threshold for minimum reads
minreadsThreshold <- 5000

# Filter out samples with read counts below the minimum threshold in bact
cat("Number of samples before filtering in bact:", nsamples(bact), "\n")
bact <- prune_samples(colSums(otu_table(bact)) > minreadsThreshold, bact)
cat("Number of samples after filtering in bact:", nsamples(bact), "\n")

# Filter out samples with read counts below the minimum threshold in fung
cat("Number of samples before filtering in fung:", nsamples(fung), "\n")
fung <- prune_samples(colSums(otu_table(fung)) > minreadsThreshold, fung)
cat("Number of samples after filtering in fung:", nsamples(fung), "\n")
```

#### 7) Alpha diversity calculations

```{r}
# Alpha diversity in bact
bact_alpha <- estimate_richness(bact, split=TRUE, measures=c("Shannon", "Observed"))
sample_data(bact)$Shannon <- bact_alpha[,"Shannon"]
sample_data(bact)$Observed <- log(bact_alpha[,"Observed"])

# Alpha diversity in fung
fung_alpha <- estimate_richness(fung, split=TRUE, measures=c("Shannon", "Observed"))
sample_data(fung)$Shannon <- fung_alpha[,"Shannon"]
sample_data(fung)$Observed <- log(fung_alpha[,"Observed"])

# Number of viruses
vir_alpha <- estimate_richness(vir, split=TRUE, measures=c("Observed"))
sample_data(vir)$Observed <- vir_alpha[,1]
```

#### 8) Datasets splitting

```{r}
# Get IDs of patients with longitudinal data
longitudinal.patients <- metadata[metadata$Dataset == "Longitudinal",]$Patient
acute.patients <- na.trim(df[df$Dataset == "1Acute" & df$keepacute == "Yes",]$Patient)
acute.samples <- na.trim(df$BCNasalswabAcute)

# Split datasets for bact
bact.long <- prune_samples(sample_data(bact)$Patient %in% longitudinal.patients, bact)
bact.long.wheeze <- prune_samples(sample_data(bact.long)$Wheeze1YEAR == "Yes", bact.long)
bact.longw1Y <- prune_samples(!(sample_names(bact) %in% acute.samples), bact)
bact.1Y <- prune_samples(sample_data(bact)$Timepoint == "T4-1year" & !(sample_names(bact) %in% acute.samples), bact)
bact.wheeze.1Y <- prune_samples(sample_data(bact.1Y)$Wheeze1YEAR == "Yes", bact.1Y)

# Split datasets for fung
fung.long <- prune_samples(sample_data(fung)$Patient %in% longitudinal.patients, fung)
fung.long.wheeze <- prune_samples(sample_data(fung.long)$Wheeze1YEAR == "Yes", fung.long)
fung.longw1Y <- prune_samples(!(sample_names(fung) %in% acute.samples), fung)
fung.1Y <- prune_samples(sample_data(fung)$Timepoint == "T4-1year" & !(sample_names(fung) %in% acute.samples), fung)
fung.wheeze.1Y <- prune_samples(sample_data(fung.1Y)$Wheeze1YEAR == "Yes", fung.1Y)

# Split datasets for vir
vir.long <- prune_samples(sample_data(vir)$Patient %in% longitudinal.patients, vir)
vir.longw1Y <- prune_samples(sample_data(vir)$Group != "Wheeze acute", vir)
vir.1Y <- prune_samples(sample_data(vir)$Timepoint == "T4-1year" & sample_data(vir)$Group != "Wheeze acute", vir)
vir.wheeze.1Y <- prune_samples(sample_data(vir.1Y)$Wheeze1YEAR == "Yes", vir.1Y)
```

# [Microbiome longitudinal investigation]{.underline}

#### 1) Alpha diversity analysis

```{r}
# Richness with age
(bact.Observed.age <- BoxplotMetadataAge(bact.long, "Observed", "Long-Group : Nasal bacterial richness with age", "Age (months)", "Observed richness (log)", custom_palette_age))
(fung.Observed.age <- BoxplotMetadataAge(fung.long, "Observed", "Long-Group : Nasal fungal richness with age", "Age (months)", "Observed richness (log)", custom_palette_age))
(vir.Observed.age <- BoxplotMetadataAge(vir.long, "Observed", "Long-Group : Nasal viral richness with age", "Age (months)", "Observed richness (log)", custom_palette_age))

# Statistics diversity
bact.Shannon.lm <- lm(SampleDataframe(bact.long)$Shannon ~ SampleDataframe(bact.long)$TimepointMonth)
summary(bact.Shannon.lm)
fung.Shannon.lm <- lm(SampleDataframe(fung.long)$Shannon ~ SampleDataframe(fung.long)$TimepointMonth)
summary(fung.Shannon.lm)

# Statistics richness 
bact.Observed.lm <- lm(SampleDataframe(bact.long)$Observed ~ SampleDataframe(bact.long)$TimepointMonth)
summary(bact.Observed.lm)
fung.Observed.lm <- lm(SampleDataframe(fung.long)$Observed ~ SampleDataframe(fung.long)$TimepointMonth)
summary(fung.Observed.lm)
fung.Observed.lm <- lm(SampleDataframe(vir.long)$Observed ~ SampleDataframe(vir.long)$TimepointMonth)
summary(fung.Observed.lm)
```

#### 2) Viruses investigation

```{r}
# Transform into any virus
taxa_sums(vir.long)
sample_data(vir.long)$AnyVirus <- ifelse(sample_sums(vir.long) == 0, "No Virus", ifelse(sample_sums(vir.long) == 1, "1 Virus detected", ">1 Viruses detected"))
table(sample_data(vir.long)$AnyVirus, sample_data(vir.long)$Group)
vir.data <- SampleDataframe(vir.long)

# Stacked bar plots of viral detection
(virus.plot.detection.long <- ggplot(vir.data %>% dplyr::count(AnyVirus, TimepointFactor) %>% mutate(pct = n / sum(n) * 100),
                               aes(TimepointFactor, pct, fill = AnyVirus)) +
  geom_bar(stat = "identity", position = "fill") +
  labs(title = "Long-Group : Nasal viral richness with age", x = "TimepointFactor", y = "Percentage of samples") +
  geom_text(aes(label = n), position = position_fill(vjust = 0.5), size = 2) +
  scale_fill_manual(values = c("#BF812D", "#DFC27D", "#F5F5F5")) +
  scale_y_continuous(labels = function(x) paste0(x*100, "%")) +
    theme(legend.position = c(1, 1), legend.box.background=element_rect(), text=element_text(size=12)))

# Detailed virus data
vir.data.detailed <- cbind(as.data.frame(taxa_sums(phyloseq::subset_samples(vir.long, TimepointFactor == "T1-1week"))),
                           as.data.frame(taxa_sums(phyloseq::subset_samples(vir.long, TimepointFactor == "T2-3months"))),
                           as.data.frame(taxa_sums(phyloseq::subset_samples(vir.long, TimepointFactor == "T3-6months"))),
                           as.data.frame(taxa_sums(phyloseq::subset_samples(vir.long, TimepointFactor == "T4-1year"))))
colnames(vir.data.detailed) <- levels(vir.data$TimepointFactor)
vir.data.detailed$Virus <- rownames(vir.data.detailed)
vir.data.detailed <- melt(vir.data.detailed)

# Bar plots of detailed virus data
(virus.plot.details.long <- ggplot(vir.data.detailed, aes(x = variable, y = value, fill = Virus)) +
  geom_bar(stat = "identity", position = "fill") +
  geom_text(aes(label = value), position = position_fill(vjust = 0.5), size = 2) +
  labs(title = "Long-Group : Nasal viral composition with age", x = "Group", y = "% of positive samples for a given virus") +
  scale_y_continuous(labels = function(x) paste0(x * 100, "%")) +
  scale_fill_brewer(palette = "BrBG") +
    theme(legend.position = c(1, 1), legend.box.background=element_rect(), text=element_text(size=12)))
```

#### 3) Beta diversity analysis

```{r}
# Normalize datasets
bact.logCSS.long <- bact.long
otu_table(bact.logCSS.long) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(bact.logCSS.long), p = 0.5), norm = TRUE, log = TRUE), taxa_are_rows = TRUE)  

fung.logCSS.long <- fung.long
otu_table(fung.logCSS.long) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(fung.logCSS.long), p = 0.5), norm = TRUE, log = TRUE), taxa_are_rows = TRUE)  

# Beta diversity
(bact.ordi.long <- PlotOrdinationAge(bact.logCSS.long, "Long-Group: Nasal bacterial weighted Unifrac PCoA", c(custom_palette_age, custom_palette_2groups)))
(fung.ordi.long <- PlotOrdinationAge(fung.logCSS.long, "Long-Group: Nasal fungal weighted Unifrac PCoA", custom_palette_age))
```

#### 4) Distance from previous timepoint

```{r}
(bact.long.distance <- PlotUnifracTimepoint(bact.logCSS.long,  "Long-Group: Bacterial paired sample distance"))
wilcox.test(UnifracTimepoint(bact.logCSS.long)$value[UnifracTimepoint(bact.logCSS.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")] ~ UnifracTimepoint(bact.logCSS.long)$variable[UnifracTimepoint(bact.logCSS.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")])

(fung.long.distance <- PlotUnifracTimepoint(fung.logCSS.long, "Long-Group: Fungal paired sample distance"))
wilcox.test(UnifracTimepoint(fung.logCSS.long)$value[UnifracTimepoint(fung.logCSS.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")] ~ UnifracTimepoint(fung.logCSS.long)$variable[UnifracTimepoint(fung.logCSS.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")])
```

#### 5) ANOSIM

```{r}
ANOSIMunifrac(bact.logCSS.long, "TimepointMonth")
ANOSIMunifrac(fung.logCSS.long, "TimepointMonth")
```

#### 6) Differential abundance testing

```{r}
# Differential abundance testing
bact.long.age.meco <- DAtesting(bact.long, c("TimepointMonth", "Patient"), "~TimepointMonth+(1|Patient)", "ASV")
table(bact.long.age.meco$res_diff$Significance, bact.long.age.meco$res_diff$Factors)
fung.long.age.meco <- DAtesting(fung.long, c("TimepointMonth", "Patient"), "~TimepointMonth+(1|Patient)", "ASV")
table(fung.long.age.meco$res_diff$Significance, fung.long.age.meco$res_diff$Factors)

# Barplot with top results
(bact.long.age.bar <- DAtestingPlotBar(bact.long.age.meco, "TimepointMonth", palette = c(custom_palette_age[1], custom_palette_age[3]), number_features = 20, title = "Long-Group : Nasal bacterial DA features with age"))
```

# [Microbiome covariates investigation]{.underline}

#### 1) Alpha diversity analysis

```{r}
# Select the covariates
covariates.env <- c("DayCare1YEAR", "NbSiblings", "Pets", "Season", "Country")
covariates.clinical <- c("AbxSinceBorn1YEAR")
covariates.genetic <- c("Gender", "MotherHayfever", "FatherHayfever", "MotherAsthma", "FatherAsthma")
covariates.birth <- c("DeliverySimplified", "GestAgeBirth", "WeightBirthKg", "BreastfeedingBirth", "AbxCoursesPregnancySimplified")
covariates.all <- c(covariates.env, covariates.clinical, covariates.genetic, covariates.birth)

# Select covariates and transform into numeric values
bact.metadata <- SampleDataframe(bact.1Y)
bact.metadata <- SelectIntoNumeric(bact.metadata, covariates.all)
bact.factors.cor <- cbind(Observed=SampleDataframe(bact.1Y)$Observed)
bact.cor <- psych::corr.test(bact.factors.cor, bact.metadata, method="spearman", adjust="fdr")

# Alpha diversity versus covariates for fung
fung.metadata <- SampleDataframe(fung.1Y)
fung.metadata <- SelectIntoNumeric(fung.metadata, covariates.all)
fung.factors.cor <- cbind(Observed=SampleDataframe(fung.1Y)$Observed)
fung.cor <- psych::corr.test(fung.factors.cor, fung.metadata, method="spearman", adjust="fdr")

# Alpha diversity versus covariates for vir
vir.metadata <- SampleDataframe(vir.1Y)
vir.metadata <- SelectIntoNumeric(vir.metadata, covariates.all)
vir.factors.cor <- cbind(Observed=SampleDataframe(vir.1Y)$Observed)
vir.cor <- psych::corr.test(vir.factors.cor, vir.metadata, method="spearman", adjust="fdr")

# Combine all correlations
all.cor <- rbind(bact.cor$r, fung.cor$r, vir.cor$r)
rownames(all.cor) <- c("Bacteria", "Fungi", "Viruses")
colnames(all.cor) <- c("Antibiotics courses during pregnancy",
                       "Any breastfeeding in the first year",
                       "Weight at birth",
                       "Gestational age at birth",
                       "Delivery mode",
                       "Father history of asthma",
                       "Mother history of asthma",
                       "Father history of hayfever",
                       "Mother history of hayfever",
                       "Sex",
                       "Antibiotics courses in the first year",
                       "More than 1 cold in the first year",
                       "Sampling season",
                       "Pets at home",
                       "Number of siblings",
                       "Daycare attendance")

# Combine all p-values
all.p <- rbind(bact.cor$p, fung.cor$p, vir.cor$p)
rownames(all.p) <-  c("Bacteria", "Fungi", "Viruses")

# Plot the correlations
(all.cor.factors <- ggcorrplot(all.cor, method="circle", p.mat=all.p, insig="blank", colors= c("goldenrod3", "white", "#964091"), outline.color = "white") +
    scale_y_discrete(labels = str_wrap(colnames(all.cor), width=30)) +
    theme(legend.position = "right", legend.box.background=element_rect(), axis.text.x=element_text(angle=30), 
          text=element_text(size=12)))

# Save plot
ggsave("Output/Figure3_1Y_Covariates_Correlations.pdf", all.cor.factors, width=35, height=20, units="cm", useDingbats=FALSE)
```

```{r}
# Add information for each virus
sample_data(vir.1Y)$Adenovirus <- ifelse(OTUDataframe(vir.1Y)["Adenovirus",]>0, "Yes", "No")
sample_data(vir.1Y)$Enterovirus <- ifelse(OTUDataframe(vir.1Y)["Enterovirus",]>0, "Yes", "No")
sample_data(vir.1Y)$Parainfluenza <- ifelse(OTUDataframe(vir.1Y)["Human Parainfluenza",]>0, "Yes", "No")
sample_data(vir.1Y)$Metapneumovirus <- ifelse(OTUDataframe(vir.1Y)["Metapneumovirus",]>0, "Yes", "No")
sample_data(vir.1Y)$Rhinovirus <- ifelse(OTUDataframe(vir.1Y)["Rhinovirus",]>0, "Yes", "No")
sample_data(vir.1Y)$RSV <- ifelse(OTUDataframe(vir.1Y)["RSV",]>0, "Yes", "No")

# Transform into any virus
taxa_sums <- taxa_sums(vir.1Y)
sample_data(vir.1Y)$AnyVirus <- ifelse(sample_sums(vir.1Y)==0, "No Virus", ifelse(sample_sums(vir.1Y)==1, "1 Virus detected", ">1 Viruses detected"))
```

```{r}
# Plot significant bacteria alpha diversity
(alpha.bact.colds <- BoxplotMetadataParameters(bact.1Y, 
                                               "Observed", 
                                               "Colds1YEARSimplified", 
                                               "CS-Group : Nasal bacterial richness with the number of colds in the first year",
                                               "Observed (log)",
                                               c(scales::alpha(custom_palette_microbes[1], 0.65),
                                                 custom_palette_microbes[1]),
                                               c(scales::alpha(custom_palette_microbes[1], 0.7),
                                                 custom_palette_microbes[1])))

(alpha.bact.abx <- BoxplotMetadataParameters(bact.1Y,
                                             "Observed",
                                             "AbxSinceBorn1YEARSimplified",
                                             "CS-Group : Nasal bacterial richness with antibiotics treatments in the first year",
                                             "Observed (log)",
                                             c(scales::alpha(custom_palette_microbes[1], 0.65),
                                               custom_palette_microbes[1]),
                                             c(scales::alpha(custom_palette_microbes[1], 0.7),
                                               custom_palette_microbes[1])))
wilcox.test(sample_data(bact.1Y)$Observed ~ sample_data(bact.1Y)$AbxSinceBorn1YEARSimplified)

(alpha.bact.daycare <- BoxplotMetadataParameters(bact.1Y,
                                                 "Observed",
                                                 "DayCare1YEAR",
                                                 "CS-Group : Bacterial Nasal richness with daycare attendance",
                                                 "Observed (log)", c(scales::alpha(custom_palette_microbes[1], 0.65),
                                                                     custom_palette_microbes[1]),
                                                 c(scales::alpha(custom_palette_microbes[1], 0.7),
                                                   custom_palette_microbes[1])))
wilcox.test(sample_data(bact.1Y)$Observed ~ sample_data(bact.1Y)$DayCare1YEAR)

(alpha.bact.siblings <- BoxplotMetadataParameters(bact.1Y,
                                                  "Observed",
                                                  "NbSiblingsSimplified",
                                                  "CS-Group : Nasal bacterial richness with siblings number",
                                                  "Observed (log)", c(scales::alpha(custom_palette_microbes[1], 0.65),
                                                                      custom_palette_microbes[1]),
                                                  c(scales::alpha(custom_palette_microbes[1], 0.7),
                                                    custom_palette_microbes[1])))

# Plot significant fungi alpha diversity
fung.season <- SampleDataframe(ps_drop_incomplete(fung.1Y, "Season"))
fung.season$Season <- as.factor(fung.season$Season)
(alpha.fung.season <- BoxplotMetadataParameters(fung.1Y,
                                                "Observed",
                                                "Season",
                                                "CS-Group : Nasal fungal diversity with sampling season",
                                                "Observed (log)",
                                                c(scales::alpha(custom_palette_microbes[2], 0.50),
                                                  scales::alpha(custom_palette_microbes[2], 0.75),
                                                  scales::alpha(custom_palette_microbes[2], 0.90), custom_palette_microbes[2]),
                                                c(scales::alpha(custom_palette_microbes[2], 0.50),
                                                  scales::alpha(custom_palette_microbes[2], 0.75),
                                                  scales::alpha(custom_palette_microbes[2], 0.90),
                                                  custom_palette_microbes[2])) +
    scale_x_discrete(labels = c(paste0(levels(fung.season$Season)[1], '\n', 'n = ', table(fung.season$Season)[1]),
                                paste0(levels(fung.season$Season)[2], '\n', 'n = ', table(fung.season$Season)[2]),
                                paste0(levels(fung.season$Season)[3], '\n', 'n = ', table(fung.season$Season)[3]),
                                paste0(levels(fung.season$Season)[4], '\n', 'n = ', table(fung.season$Season)[4]))) +
    geom_signif(comparisons=list(c("1-Winter", "3-Summer")),
                map_signif_level=c("***"=0.001/6, "**"=0.01/6, "*"=0.05/6),  y_position=c(4.5, 4)))

# Plot significant viruses alpha diversity
(alpha.vir.daycare <- plotanyViruses(vir.1Y, "DayCare1YEAR",
                                     "CS-Group : Nasal viral richness with daycare attendance"))
vir.1Y.noNA.daycare <- ps_drop_incomplete(vir.1Y, vars = "DayCare1YEAR", verbose = TRUE)
wilcox.test(sample_data(vir.1Y.noNA.daycare)$Observed ~ sample_data(vir.1Y.noNA.daycare)$DayCare1YEAR)

(alpha.vir.abx <- plotanyViruses(vir.1Y, "AbxSinceBorn1YEARSimplified", 
                                 "CS-Group : Nasal viral richness with antibiotics treatments in the first year"))
wilcox.test(sample_data(vir.1Y)$Observed ~ sample_data(vir.1Y)$AbxSinceBorn1YEARSimplified)
```

#### 2) dbRDA variables selection

```{r}
# Set covariates of interest
covariates.simplified.env <- c("DayCare1YEAR", "NbSiblingsSimplified", "Pets", "Season")
covariates.simplified.clinical <- c("Colds1YEARSimplified", "AbxSinceBorn1YEARSimplified")
covariates.simplified.genetic <- c("Gender", "MotherHayfever", "FatherHayfever", "MotherAsthma", "FatherAsthma")
covariates.simplified.birth <- c("DeliverySimplified", "GestAgeBirthSimplified", "WeightBirthKgSimplified", "BreastfeedingBirth", "AbxCoursesPregnancySimplified")
covariates.simplified.all <- c(covariates.simplified.env, covariates.simplified.clinical, covariates.simplified.genetic, covariates.simplified.birth)

# Normalize datasets
bact.logCSS.1Y <- bact.1Y
otu_table(bact.logCSS.1Y) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(bact.logCSS.1Y), p = 0.5), norm = TRUE, log = TRUE), taxa_are_rows = TRUE)  
fung.logCSS.1Y <- fung.1Y
otu_table(fung.logCSS.1Y) <- otu_table(MRcounts(cumNorm(phyloseq_to_metagenomeSeq(fung.logCSS.1Y), p = 0.5), norm = TRUE, log = TRUE), taxa_are_rows = TRUE)  

# Select variables based on dbRDA
bact.1Y.dbrda <- dbRDAselectionUnifrac(bact.logCSS.1Y, covariates.simplified.all)
fung.1Y.dbrda <- dbRDAselectionUnifrac(fung.logCSS.1Y, covariates.simplified.all)
bact.1Y.dbrda
fung.1Y.dbrda
```

#### 3) Differential abundance testing

```{r}
# Calculate the number of DA ASVs for each covariate
bact.1Y.covariates <- DAfeaturesCounts(bact.1Y, covariates.simplified.all, "ASV")
fung.1Y.covariates <- DAfeaturesCounts(fung.1Y, covariates.simplified.all, "ASV")

# New variable names
newnames <- c("Antibiotics courses during pregnancy", "Any breastfeeding in the first year",
              "Weight at birth", "Gestational age at birth", "Delivery mode", 
              "Father history of asthma",  "Mother history of asthma", 
              "Father history of hayfever", "Mother history of hayfever",
              "Sex", "Antibiotics courses in the first year",
              "More than 2 colds in the first year", "Sampling season",
              "Pets at home", "Number of siblings",
              "Daycare attendance in the first year")

# Results and reordering
bact.1Y.covariates$Newname <- rev(newnames)
fung.1Y.covariates$Newname <- rev(newnames)

# Plot
(covariates.da.bact.plot <- PlotfeaturesCounts(bact.1Y.covariates,
                                               custom_palette_omics2[3],
                                               "CS-Group : Nasal bacterial DA features with covariates"))
(covariates.da.fung.plot <- PlotfeaturesCounts(fung.1Y.covariates,
                                               custom_palette_omics2[4],
                                               "CS-Group : Nasal fungal DA features with covariates"))

# Colds variable
bact.da.colds <- DAtesting(bact.1Y, "Colds1YEARSimplified", "~Colds1YEARSimplified", "ASV")
(bact.da.colds.bar <- DAtestingPlotBar(bact.da.colds,
                                       "More than 1 colds in the first year - 1 or less colds in the first year",
                                       rep(custom_palette_microbes[1],2), 7,
                                       "CS-Group : Nasal bacterial DA features with colds in the first year"))

# Day care variable
bact.da.daycare <- DAtesting(bact.1Y, "DayCare1YEAR", "~DayCare1YEAR", "ASV")
(bact.da.daycare.bar <- DAtestingPlotBar(bact.da.daycare,
                                         "Yes daycare - No daycare",
                                         rep(custom_palette_microbes[1],2), 19,
                                         "CS-Group : Nasal bacterial DA features with daycare attendance"))

# Siblings variable
bact.da.siblings <- DAtesting(bact.1Y, "NbSiblingsSimplified", "~NbSiblingsSimplified", "ASV")
(bact.da.siblings.bar <- DAtestingPlotBar(bact.da.siblings,
                                          "One or more sibling(s) - No siblings",
                                          rep(custom_palette_microbes[1],2), 7,
                                          "CS-Group : Nasal bacterial DA features with siblings number"))

# Season
fung.da.season <- DAtesting(fung.1Y, "Season", "~Season", "ASV")
table(fung.da.season$res_diff$Comparison, fung.da.season$res_diff$Significance)
(fung.da.season.bar <- DAtestingPlotBar(fung.da.season,
                                        c("3-Summer - 1-Winter",  "4-Autumn - 1-Winter"),
                                        c(custom_palette_microbes), 14,
                                        "CS-Group : Nasal fungal DA features with sampling season"))
```

# [Microbiome disease investigation]{.underline}

#### 1) Alpha diversity analysis

```{r}
# Bacterial and fungal diversity and richness for 1 year dataset
(bact.Shannon.1Y <- BoxplotMetadataGroup(bact.1Y, "Shannon", "CS-Group : Nasal bacterial diversity with wheeze", "Shannon diversity", custom_palette_2groups))
(bact.Observed.1Y <- BoxplotMetadataGroup(bact.1Y, "Observed", "CS-Group : Nasal bacterial richness with wheeze", "Observed richness (log)", custom_palette_2groups))
(fung.Shannon.1Y <- BoxplotMetadataGroup(fung.1Y, "Shannon", "CS-Group : Nasal fungal diversity with wheeze", "Shannon diversity", custom_palette_2groups))
(fung.Observed.1Y <- BoxplotMetadataGroup(fung.1Y, "Observed", "CS-Group : Nasal fungal richness with wheeze", "Observed richness (log)", custom_palette_2groups))
(vir.Observed.1Y <- BoxplotMetadataGroup(vir.1Y, "Observed", "CS-Group : Nasal viral richness with wheeze", "Observed richness (number of detected viruses)", custom_palette_2groups))

# Statistical tests
wilcox.test(SampleDataframe(bact.1Y)$Shannon ~ SampleDataframe(bact.1Y)$Wheeze1YEAR)
wilcox.test(SampleDataframe(bact.1Y)$Observed ~ SampleDataframe(bact.1Y)$Wheeze1YEAR)
wilcox.test(SampleDataframe(fung.1Y)$Shannon ~ SampleDataframe(fung.1Y)$Wheeze1YEAR)
wilcox.test(SampleDataframe(fung.1Y)$Observed ~ SampleDataframe(fung.1Y)$Wheeze1YEAR)
```

#### 2) Viruses investigation

```{r}
# Run correlation tests
cor.test(sample_data(vir.1Y)$Observed, as.numeric(as.factor(sample_data(vir.1Y)$Group)), method = "spearman")

# Run proportion tests
prop.test(table(sample_data(vir.1Y)$AnyVirus, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$Adenovirus, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$Enterovirus, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$Parainfluenza, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$Metapneumovirus, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$Rhinovirus, sample_data(vir.1Y)$Group))
prop.test(table(sample_data(vir.1Y)$RSV, sample_data(vir.1Y)$Group))

# Convert to data frame
vir.data <- as.data.frame(SampleDataframe(vir.1Y))

# Stacked bar plots of viral detection
(virus.plot.wheeze <- plotanyViruses(vir.1Y, "Group", "CS-Group : Nasal viral richness with wheeze"))

# Combine virus data for each group
vir.data.detailed <- cbind(as.data.frame(taxa_sums(phyloseq::subset_samples(vir.1Y, Group=="No"))),
                            as.data.frame(taxa_sums(phyloseq::subset_samples(vir.1Y, Group=="Yes"))))
colnames(vir.data.detailed) <- c("Ctrl", "Wheeze")
vir.data.detailed$Virus <- rownames(vir.data.detailed)
vir.data.detailed <- melt(vir.data.detailed)

# Plot positive viruses by group
(virus.plot.details.wheeze <- ggplot(vir.data.detailed, aes(x = variable, y = value, fill = Virus)) +
    geom_bar(stat = "identity", position = "fill") +
    geom_text(aes(label = value), position = position_fill(vjust = 0.5), size = 5) +
    labs(title = "CS-Group : Nasal viral composition with wheeze", x = "Group", y = "% of positive samples for a given virus") +
    scale_y_continuous(labels = function(x) paste0(x * 100, "%")) +
    scale_x_discrete(labels = c(paste0("Controls", '\n', 'n = ', table(SampleDataframe(vir.1Y)$Group)[1]),
                             paste0("Wheezers", '\n', 'n = ', table(SampleDataframe(vir.1Y)$Group)[2]))) +
    scale_fill_brewer(palette = "BrBG") +
    theme(legend.position = c(1,0),
          legend.box.background = element_rect(),
          text = element_text(size = 12),
          legend.title = element_blank(),
          legend.text = element_text(size = 10),
          legend.key.size = unit(0.5, "cm")) +
    guides(alpha = "none"))
```

#### 3) Differential abundance testing

```{r}
# Differential abundance testing bacteria
bact.da.disease.ASV.adj <- DAtesting(bact.1Y, c("Group", "DayCare1YEAR"),"~Group+DayCare1YEAR", level="ASV")
table(bact.da.disease.ASV.adj$res_diff$Factors, bact.da.disease.ASV.adj$res_diff$Significance)
bact.da.disease.ASV <- DAtesting(bact.1Y, "Group","~Group", level="ASV")
table(bact.da.disease.ASV$res_diff$Comparison, bact.da.disease.ASV$res_diff$Significance)
(bact.da.disease.adj <- DAtestingPlotBar(bact.da.disease.ASV.adj, 
                                         "GroupYes", 
                                         custom_palette_2groups, 7, 
                                         "CS-Group : Nasal bacterial DA features with wheeze (adjusted for daycare attendance)"))
(bact.da.disease <- DAtestingPlotBar(bact.da.disease.ASV,
                                     "Yes - No",
                                     custom_palette_2groups, 14,
                                     "CS-Group : Nasal bacterial DA features with wheeze"))

# Differential abundance testing fungi
fung.da.disease.ASV.adj <- DAtesting(fung.1Y, c("Group", "Season"),"~Group+(1|Season)", level="ASV")
table(fung.da.disease.ASV.adj$res_diff$Factors, fung.da.disease.ASV.adj$res_diff$Significance)
fung.da.disease.ASV <- DAtesting(fung.1Y, "Group","~Group", level="ASV")
table(fung.da.disease.ASV$res_diff$Comparison, fung.da.disease.ASV$res_diff$Significance)
```

# Microbiome commonalities between the cohorts

#### 1) Differential abundance testing at each timepoint

```{r}
# Number of samples at each time point for which we have wheeze start month
table(sample_data(bact.long.wheeze)$WheezeStartMonth, sample_data(bact.long.wheeze)$TimepointSimplified)
summary(sample_data(bact.long.wheeze)$WheezeStartMonth)

# DA testing on total model
bact.long.group <- DAtesting(bact.long, "Group", "~Group+TimepointMonth+(1|Patient)", level="ASV")
table(bact.long.group$res_diff$Factors, bact.long.group$res_diff$Significance)
```

#### 2) Haemophilus detection

```{r}
# Add "Haemophilus_influenzae_19 levels
sample_data(bact.logCSS.long)$Haemophilus <- OTUDataframe(bact.logCSS.long)[taxa_names(bact.logCSS.long)=="Haemophilus_influenzae_19",]
sample_data(bact.logCSS.long)$DetectedHaemophilus <- ifelse(sample_data(bact.logCSS.long)$Haemophilus>0, TRUE, FALSE)

# Create dataframe for survival analysis
survival.df <- data.frame(
  time = sample_data(bact.logCSS.long)$TimepointMonth,
  daycare = sample_data(bact.logCSS.long)$DayCare1YEAR,
  haemophilus = sample_data(bact.logCSS.long)$DetectedHaemophilus,
  group = sample_data(bact.logCSS.long)$Group
)

# Perform survival analysis
fitHaemophilus <- survfit(Surv(time, haemophilus) ~ group, data = survival.df)
summary(fitHaemophilus)
survdiff(Surv(time, haemophilus) ~ group, data = survival.df)

# Survival plots
(Haemophilus.survival.plot <- ggsurvplot(fitHaemophilus, fun = "event", pval = TRUE, conf.int = TRUE, risk.table = FALSE, surv.median.line = "none", palette = custom_palette_2groups)) +
    xlab("Age (months)") + ylab("Proportion of samples within group") +
  ggtitle("Long-Group : Haemophilus_influenzae_19 positive patients survival analysis with wheeze")
```

# [Host data import]{.underline}

#### 1) Metadata import

```{r}
# Load metadata
source("Archive/Metadata_host_nasal.R") # REMOVE
source("Archive/Metadata_host_blood.R") # REMOVE
saveRDS(df.nasal, "Input/df.nasal.rds")
saveRDS(df.blood.1Y, "Input/df.blood.1Y.rds")
df.nasal <- readRDS("Input/df.nasal.rds")
df.blood.1Y <- readRDS("Input/df.blood.1Y.rds")
```

#### 2) Gene expression data import

```{r}
# Load RNAseq object
identical(rownames(host.nasal), rownames(host.blood))

# Get Ensembl annotations
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
ensemblsIDS <- rownames(host.nasal)
list <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
              filters = "ensembl_gene_id", values = ensemblsIDS, mart = ensembl)
colnames(list) <- c("gene_id", "hgnc_symbol", "gene_biotype")

# Add gene information
rowData(host.nasal) <- merge(list, rowData(host.nasal), by = "gene_id")
rowData(host.blood) <- merge(list, rowData(host.blood), by = "gene_id")

# Check metadata matching
identical(rownames(df.nasal), colnames(host.nasal))
identical(rownames(df.blood.1Y), colnames(host.blood))

# Create DGEList object
host.dge.nasal <- DGEList(counts=assays(host.nasal)$counts, samples=df.nasal, group=df.nasal$Group, genes=rowData(host.nasal), remove.zeros=TRUE)
host.dge.blood <- DGEList(counts=assays(host.blood)$counts, samples=df.blood.1Y, group=df.blood.1Y$Group, genes=rowData(host.blood), remove.zeros=TRUE)
```

# [Host data pre-processing]{.underline}

#### 1) Filtering and normalization

```{r}
# Remove samples with less than 10Mio mapped reads
dim(host.dge.nasal)
host.dge.nasal <- host.dge.nasal[,host.dge.nasal$samples$lib.size>10e6]
dim(host.dge.nasal)
dim(host.dge.blood)
host.dge.blood <- host.dge.blood[,host.dge.blood$samples$lib.size>10e6]
dim(host.dge.blood)

# Normalize and run voom transformation
host.dge.nasal <- edgeR::calcNormFactors(host.dge.nasal)
host.dge.blood <- edgeR::calcNormFactors(host.dge.blood)

# Filtering using the design information
design.nasal <- model.matrix(~Group, data = host.dge.nasal$samples)
keep.nasal <- filterByExpr(host.dge.nasal, design.nasal)
host.dge.nasal <- host.dge.nasal[keep.nasal, ]
dim(host.dge.nasal)
design.blood <- model.matrix(~Group, data = host.dge.blood$samples)
keep.blood <- filterByExpr(host.dge.blood, design.blood)
host.dge.blood <- host.dge.blood[keep.blood, ]
dim(host.dge.blood)

# Voom transformation
host.dge.nasal <- voom(host.dge.nasal, design.nasal, plot = TRUE)
host.dge.blood <- voom(host.dge.blood, design.blood, plot = TRUE)

# Select protein coding defined genes only
host.dge.nasal <- host.dge.nasal[host.dge.nasal$genes$gene_biotype == "protein_coding" & host.dge.nasal$genes$hgnc_symbol != "", ]
dim(host.dge.nasal)
host.dge.blood <- host.dge.blood[host.dge.blood$genes$gene_biotype == "protein_coding" & host.dge.blood$genes$hgnc_symbol != "", ]
dim(host.dge.blood)

# Select immune genes only
ImmuneDB <- read.table("Input/GeneList.txt", sep = "\t", header = TRUE)
ImmuneDB$name <- ImmuneDB$Symbol
host.dge.nasal$genes$immune <- ifelse(host.dge.nasal$genes$hgnc_symbol %in% ImmuneDB$name, "Yes", "No")
host.dge.blood$genes$immune <- ifelse(host.dge.blood$genes$hgnc_symbol %in% ImmuneDB$name, "Yes", "No")
dim(host.dge.nasal)
dim(host.dge.blood)

# Add symbol as rowname
rownames(host.dge.nasal) <- host.dge.nasal$genes$gene_name
rownames(host.dge.blood) <- host.dge.blood$genes$gene_name
```

```{r}
# Perform PCA before batch correction
logCPM.blood <- host.dge.blood$E
pca_before <- prcomp(t(logCPM.blood), scale. = TRUE)

# Create a data frame with PCA results before correction
pca_df_before <- data.frame(
  Sample = rownames(pca_before$x),
  PC1 = pca_before$x[, 1],
  PC2 = pca_before$x[, 2],
  Group = host.dge.blood$targets$group,
  Batch = host.dge.blood$targets$SeqID
)

# Plot PCA before batch correction
pca_plot_before <- ggplot(pca_df_before, aes(x = PC1, y = PC2, color = Batch, shape = Group)) +
  geom_point(size = 3) +
  labs(title = "PCA Before Batch Correction", x = "PC1", y = "PC2") +
  theme_minimal()
print(pca_plot_before)

# Perform batch correction
logCPM.blood.corrected <- removeBatchEffect(logCPM.blood, batch=host.dge.blood$targets$SeqID, design=design.blood)

# Perform PCA on batch-corrected data
pca_after <- prcomp(t(logCPM.blood.corrected), scale. = TRUE)

# Create a data frame with PCA results after correction
pca_df_after <- data.frame(
  Sample = rownames(pca_after$x),
  PC1 = pca_after$x[, 1],
  PC2 = pca_after$x[, 2],
  Group = host.dge.blood$targets$group,
  Batch = host.dge.blood$targets$SeqID
)

# Plot PCA after batch correction
pca_plot_after <- ggplot(pca_df_after, aes(x = PC1, y = PC2, color = Batch, shape = Group)) +
  geom_point(size = 3) +
  labs(title = "PCA After Batch Correction", x = "PC1", y = "PC2") +
  theme_minimal()
print(pca_plot_after)

# Update the DGEList object with batch-corrected data
host.dge.blood$E <- logCPM.blood.corrected
```

#### 2) Grouping and clinical covariates

```{r}
# Get IDs of patients with longitudinal data
longitudinal.patients <- metadata[metadata$Dataset=="Longitudinal",]$Patient
acute.patients <- na.trim(df.nasal[df.nasal$Dataset=="1Acute" & df.nasal$keepacute=="Yes",]$Patient)
acute.samples <- na.trim(df.nasal$BCNasalbrushingmRNAAcute)

# Split datasets
host.nasal.long <- host.dge.nasal[,host.dge.nasal$targets$Patient %in% longitudinal.patients]
table(host.nasal.long$targets$Timepoint)

host.nasal.longw1Y <- host.dge.nasal[,!(colnames(host.dge.nasal) %in% acute.samples)]
table(host.nasal.longw1Y$targets$Group)
table(host.nasal.longw1Y$targets$Timepoint)

host.nasal.1Y <- host.dge.nasal[,host.dge.nasal$targets$Timepoint=="T4-1year" & !(colnames(host.dge.nasal) %in% acute.samples)]
table(host.nasal.1Y$targets$Group)

host.nasal.wheeze.1Y <- host.nasal.1Y[,host.nasal.1Y$targets$Group=="Yes" & !(colnames(host.nasal.1Y) %in% acute.samples)]
host.blood.wheeze.1Y <- host.dge.blood[,host.dge.blood$targets$Group=="Yes" & !(colnames(host.dge.blood) %in% acute.samples)]
```

# [Host longitudinal investigation]{.underline}

#### 1) Principal coordinate analysis : Nasal

```{r}
# Perform PCA
PCA <- prcomp(t(host.nasal.long$E), center=TRUE)
varexp <- round(summary(PCA)$importance[2,1:2] * 100, digits = 0)

# Create dataset
data.PCA <- cbind(data.frame(Samples = rownames(PCA$x), PC1 = PCA$x[,1], PC2 = PCA$x[,2]), host.nasal.long$targets)

# Plot results
(plot.nasal.PCA.Age <- ggplot(data.PCA, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(fill = TimepointFactor), geom = "polygon", type = "t", level = 0.75, alpha = 0.2) +
  scale_shape_identity() +
  geom_xsidedensity(aes(y = after_stat(density), fill = TimepointFactor, color = TimepointFactor), size = 0.25) +
  geom_ysidedensity(aes(x = after_stat(density), fill = TimepointFactor, color = TimepointFactor), size = 0.25) +
  geom_point(aes(fill = TimepointFactor, color = TimepointFactor),
             shape = 21,
             #shape = ifelse(data.PCA$Group == "Yes" & data.PCA$TimepointFactor == "T4-1year", 25, 21),
             size = 2, stroke = 0.25) +
  annotate('text', x = min(data.PCA$PC1), y = min(data.PCA$PC2),
           label = paste0('Controls n = ', table(data.PCA$Group)[1], '\n',
                          'Wheezers n = ', table(data.PCA$Group)[2])) +
  scale_color_manual(values = adjustcolor(custom_palette_age, alpha = 1)) +
  scale_fill_manual(values = adjustcolor(custom_palette_age, alpha = 0.5)) +
  xlab(paste0("PC1 ", varexp[1], "%")) + ylab(paste0("PC2 ", varexp[2], "%")) +
  ggtitle("Long-Group : Nasal host gene expression PCA") +
  theme(ggside.panel.scale = 1/4,
        axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank(),
        legend.position = c(1, 0), legend.box.background=element_rect(), text=element_text(size=12)))

# Adonis 
adonis2(formula = as.matrix(dist(t(host.nasal.long$E), method='maximum')) ~ TimepointSimplified, data = host.nasal.long$targets)
```

#### 2) Distance from previous timepoint

```{r}
(host.long.distance <- PlotDistanceTimepoint(host.nasal.long,  "Long-Group: Nasal host gene expression paired sample distance"))
wilcox.test(DistanceTimepoint(host.nasal.long)$value[DistanceTimepoint(host.nasal.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")] ~ DistanceTimepoint(host.nasal.long)$variable[DistanceTimepoint(host.nasal.long)$variable %in% c("T1.1week.3months", "T2.3months.6months")])
```

#### 3) Differential gene expression analysis

```{r}
# Set parameters
FDR_param <- 0.1
FC_param <- 0

# Choose design
design <- model.matrix(~TimepointMonth, data=host.nasal.long$targets)
colnames(design)
corfit <- duplicateCorrelation(host.nasal.long$E, design, block=host.nasal.long$targets$Patient)

# Fit model and perform test
fit <- lmFit(host.nasal.long$E, design, correlation=corfit$consensus)
fit <- eBayes(fit)

# Get DE genes with age
DE_Age <- topTable(fit, coef=2, number="Inf")
DE_Age$Direction <- ifelse(DE_Age$logFC>FC_param & DE_Age$adj.P.Val<FDR_param, "Increase",
                           ifelse(-(DE_Age$logFC)>FC_param & DE_Age$adj.P.Val<FDR_param, "Decrease", ""))
table(DE_Age$Direction)

# Subset into immune genes and add labels
DE_Age$Immune <- ifelse(DE_Age$ID %in% ImmuneDB$name, "Yes", "No")
DE_Age$label <- NA
DE_Age$label <- ifelse(DE_Age$Direction!="" & DE_Age$Immune=="Yes", DE_Age$ID, NA)
table(DE_Age$Immune, DE_Age$Direction)

# Plot volcano
(plot.volcano.Age <- PlotVolcanoGenes(DE_Age, "Decrease", c("grey", custom_palette_age[1], custom_palette_age[3]), "Long-Group : Nasal host DE testing with age"))

# Save results to excel sheet
write.xlsx(DE_Age, "Output/Age_DE_results.xlsx", row.names=FALSE)

```

#### 4) Pathway analysis of DE genes

```{r}
immune.pathway.age <- ImmunePathwayAnalysis(DE_Age)
(plot.immune.pathway.age <- PlotImmunePathways(immune.pathway.age, palette = c(custom_palette_age[1], custom_palette_age[3]), number_pathways = 20, min_genecounts = 4, title = "Long-Group : Nasal host immune pathways with age"))

# Save pathway analysis results
write.xlsx(immune.pathway.age, "Output/Supplementary_table_1.xlsx", row.names=FALSE)

# Average pathways in relation to age
(pathways.long.age <- PlotTopPathways(host.nasal.long, immune.pathway.age, n_top_pathways = 20))

# Average pathways in relation to age with individual genes
# Pathway plots
(pathways.plot.jakstatdown <- PlotTopPathwayswGenes(host.nasal.long, immune.pathway.age, c(16), 
                                               c(custom_palette_age[1]), 
                                               c(c("JAK1", "STAT1", "STAT3", "IL15", "IL15RA", "IL6ST", "IL7R", "IL3RA", "IL5RA", "CSF2"))) +
    ylab("Average gene expression (normalised)"))
(pathways.plot.jakstatupn <- PlotTopPathwayswGenes(host.nasal.long, immune.pathway.age, c(4), 
                                               c(custom_palette_age[3]), 
                                               c("JAK2", "TYK2", "IFNG", "IFNK", "IFNA6", "IFNL3", "IL2RA", "IL21R", "IL4R", "IL10RA", "IL12RB1")))
(pathways.plot.th17 <- PlotTopPathwayswGenes(host.nasal.long, immune.pathway.age, c(7), 
                                               c(custom_palette_age[3]), 
                                               c(c("TGFBR2", "IL21R"))))
(pathways.plot.Bcell <- PlotTopPathwayswGenes(host.nasal.long, immune.pathway.age, c(8), 
                                               c(custom_palette_age[3]), 
                                               c(c("CD79A", "CD79B", "BTK"))))
(pathways.plot.Th1Th2 <- PlotTopPathwayswGenes(host.nasal.long, immune.pathway.age, c(9), 
                                               c(custom_palette_age[3]), 
                                               c(c( "IFNG", "IL12RB1", "IL2RA", "IL4R"))))
```

# [Host covariates investigation]{.underline}

#### 1) Number of DE features with covariates

```{r}
# Calculate the number of DE features
host.nasal.covariates.counts <- DEfeaturesCounts(host.nasal.1Y, covariates.all)
host.nasal.covariates.counts
host.blood.covariates.counts <- DEfeaturesCounts(host.dge.blood, covariates.all)
host.blood.covariates.counts

# New variable names
newnames <- c("Antibiotics courses during pregnancy", "Any breastfeeding in the first year",
              "Weight at birth", "Gestational age at birth", "Delivery mode", 
              "Father history of asthma",  "Mother history of asthma", "Father history of hayfever", "Mother history of hayfever",
              "Sex", "Antibiotics courses in the first year", "More than 2 colds in the first year",
              "Sampling season", "Pets at home", "Number of siblings", "Daycare attendance in the first year")

# Results and reordering
host.nasal.covariates.counts$Newname <- rev(newnames)
host.blood.covariates.counts$Newname <- rev(newnames)

# Plot number of features
(host.nasal.covariates.plot <- PlotfeaturesCounts(host.nasal.covariates.counts, custom_palette_omics2[1],"MaCo : Nasal host DE features with covariates"))
(host.blood.covariates.plot <- PlotfeaturesCounts(host.blood.covariates.counts, custom_palette_omics2[2],"MaCo : Blood host DE features with covariates"))

# Genes with daycare
host.nasal.daycare <- host.nasal.1Y[,which(!is.na(host.nasal.1Y$targets$DayCare1YEAR))]
host.nasal.daycare.results <- topTable(eBayes(lmFit(host.nasal.daycare$E, model.matrix(formula(paste0("~", "DayCare1YEAR")), data=host.nasal.daycare$targets))), coef=2, number="Inf")
head(host.nasal.daycare.results)
```

#### 2) dbRDA variables selection

```{r}
host.nasal.dbrda <- dbRDAselectionMaximum(host.nasal.1Y, covariates.simplified.all)
host.blood.dbrda <- dbRDAselectionMaximum(host.dge.blood, covariates.simplified.all)
host.nasal.dbrda
host.blood.dbrda
```

# [Host disease investigation]{.underline}

#### 1) Principal coordinate analysis : Nasal

```{r}
# Perform PCA.nasal
PCA.nasal <- prcomp(t(host.nasal.1Y$E), center=TRUE)
varexp <- c(summary(PCA.nasal)$importance[2, 1] * 100, summary(PCA.nasal)$importance[2, 2] * 100)
varexp <- round(varexp, digits = 0)

# Create dataset
data.PCA.nasal <- cbind(data.frame(Samples = rownames(PCA.nasal$x), PC1 = PCA.nasal$x[, 1], PC2 = PCA.nasal$x[, 2]), host.nasal.1Y$targets)

# Plot results
(plot.PCA.nasal.Disease <- ggplot(data.PCA.nasal, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(fill = Group), alpha = 0.2, geom = "polygon", type = "t", level = 0.75) + scale_shape_identity() +
  geom_xsidedensity(aes(y = after_stat(density), fill = Group, color = Group), size = 0.25) +
  geom_ysidedensity(aes(x = after_stat(density), fill = Group, color = Group), size = 0.25) +
  geom_point(aes(fill = Group, color = Group), shape = 21, size = 2, stroke = 0.25) +
  scale_color_manual(values = adjustcolor(custom_palette_2groups, alpha = 1)) +
  annotate('text', x=min(data.PCA.nasal$PC1), y=min(data.PCA.nasal$PC2),
             label=paste0('Controls n = ', table(data.PCA.nasal$Group)[1], '\n', 'Wheezers n = ', table(data.PCA.nasal$Group)[2])) +
  scale_fill_manual(values = adjustcolor(custom_palette_2groups, alpha = 0.5)) +
  xlab(paste0("PC1 ", varexp[1], "%")) + ylab(paste0("PC2 ", varexp[2], "%")) + ggtitle("CS-Group : Nasal host PCA") +
  theme(axis.text.x=element_blank(), axis.text.y=element_blank(), axis.ticks=element_blank(),
          legend.position = "none", legend.box.background=element_rect(), text=element_text(size=12)))

# Adonis 
adonis2(formula = as.matrix(dist(t(host.nasal.1Y$E), method='maximum')) ~ Group, data = host.nasal.1Y$targets)
```

#### 2) Principal coordinate analysis : Blood

```{r}
# Perform PCA.blood
PCA.blood <- prcomp(t(host.dge.blood$E), center=TRUE)
varexp <- c(summary(PCA.blood)$importance[2, 1] * 100, summary(PCA.blood)$importance[2, 2] * 100)
varexp <- round(varexp, digits = 0)

# Create dataset
data.PCA.blood <- cbind(data.frame(Samples = rownames(PCA.blood$x), PC1 = PCA.blood$x[, 1], PC2 = PCA.blood$x[, 2]), host.dge.blood$targets)

# Plot results
(plot.PCA.blood.Disease <- ggplot(data.PCA.blood, aes(x = PC1, y = PC2)) +
  stat_ellipse(aes(fill = Group), alpha = 0.2, geom = "polygon", type = "t", level = 0.75) + scale_shape_identity() +
  geom_xsidedensity(aes(y = after_stat(density), fill = Group, color = Group), size = 0.25) +
  geom_ysidedensity(aes(x = after_stat(density), fill = Group, color = Group), size = 0.25) +
  geom_point(aes(fill = Group, color = Group), shape = 21, size = 2, stroke = 0.25) +
  scale_color_manual(values = adjustcolor(custom_palette_2groups, alpha = 1)) +
  annotate('text', x=min(data.PCA.blood$PC1), y=min(data.PCA.blood$PC2),
             label=paste0('Controls n = ', table(data.PCA.blood$Group)[1], '\n', 'Wheezers n = ', table(data.PCA.blood$Group)[2])) +
  scale_fill_manual(values = adjustcolor(custom_palette_2groups, alpha = 0.5)) +
  xlab(paste0("PC1 ", varexp[1], "%")) + ylab(paste0("PC2 ", varexp[2], "%")) + ggtitle("MaCo : Blood host PCA") +
  theme(axis.text.x=element_blank(), axis.text.y=element_blank(), axis.ticks=element_blank(),
          legend.position = "none", legend.box.background=element_rect(), text=element_text(size=12)))

# Adonis
adonis2(formula = as.matrix(dist(t(host.dge.blood$E), method='maximum')) ~ Group, data = host.dge.blood$targets)
```

#### 3) Differential gene expression analysis : Nasal

```{r}
# Choose design
design.nasal <- model.matrix(~Group+Gender, data=host.nasal.1Y$targets)
colnames(design.nasal)

# Get DE genes with age
DE_Group_nasal <- topTable(eBayes(lmFit(host.nasal.1Y$E, design.nasal)), coef=2, number="Inf")
DE_Group_nasal$Direction <- ifelse(DE_Group_nasal$logFC>FC_param & DE_Group_nasal$adj.P.Val<FDR_param, "Wheeze",
                           ifelse(-(DE_Group_nasal$logFC)>FC_param & DE_Group_nasal$adj.P.Val<FDR_param, "Control", ""))
table(DE_Group_nasal$Direction)

# Add immune information
DE_Group_nasal$Immune <- ifelse(DE_Group_nasal$ID %in% ImmuneDB$name, "Yes", "No")

# Add labels for significant genes
DE_Group_nasal$label <- NA
DE_Group_nasal$label <- ifelse(DE_Group_nasal$Direction!="" & DE_Group_nasal$Immune=="Yes", DE_Group_nasal$ID, NA)
table(DE_Group_nasal$Direction, DE_Group_nasal$Immune)

# Plot volcano
(plot.volcano.nasal.Disease <- PlotVolcanoGenes(DE_Group_nasal, "Control", c("grey", custom_palette_2groups), "MaCo : Nasal host DE genes with wheeze (adjusted for sex)"))

# Save results to excel sheet
write.xlsx(DE_Group_nasal[1:500,], "Output/Group_DE_nasal_results.xlsx", row.names=FALSE)
```

#### 4) Differential gene expression analysis : Blood

```{r}
# Choose design
design.blood <- model.matrix(~Group+Gender, data=host.dge.blood$targets)
colnames(design.blood)

# Get DE genes with age
DE_Group_blood <- topTable(eBayes(lmFit(host.dge.blood$E, design.blood)), coef=2, number="Inf")
DE_Group_blood$Direction <- ifelse(DE_Group_blood$logFC>FC_param & DE_Group_blood$adj.P.Val<FDR_param, "Wheeze",
                           ifelse(-(DE_Group_blood$logFC)>FC_param & DE_Group_blood$adj.P.Val<FDR_param, "Control", ""))
table(DE_Group_blood$Direction)

# Save results to excel sheet
write.xlsx(DE_Group_blood, "Output/Group_DE_blood_results.xlsx", row.names=FALSE)
```

#### 5) Pathway analysis of DE genes

```{r}
immune.pathway.disease <- ImmunePathwayAnalysis(DE_Group_nasal)

# Save pathway analysis results
write.xlsx(immune.pathway.disease, "Output/Group_immune_pathways.xlsx", row.names=FALSE)

(immune.pathway.disease.plot <- PlotImmunePathways(immune.pathway.disease, palette = custom_palette_2groups[2], number_pathways = 2, min_genecounts = 2, title = "CS-Group : Nasal host immune pathways with wheeze (adjusted for sex)"))
```

# [Multi-omics integration]{.underline}

#### 1) Datasets preparation

```{r}
# Save a version of metadata with subject number as row name
metaS <- metadata
rownames(metaS) <- paste0("S", metaS$Subjectnumber)

# Keep metadata of interest
metaS.subset <- metaS[, colnames(metaS) %in% c(covariates.all, "Group", "DayCareGroup")]

# Set parameters for microbial data filtering and transformation
prev.omics <- 0.1
count.omics <- 1

# Filter and transform bacterial data
bact.omics <- filter_taxa(bact.1Y, function(x) sum(x>count.omics) > (prev.omics*length(x)), TRUE)
otu_table(bact.omics) <- microbiome::transform(otu_table(bact.omics), transform="clr")  
hist(as.vector(OTUDataframe(bact.omics)), breaks = 200, col = "#5C1E91")

# Filter and transform fungal data
fung.omics <- filter_taxa(fung.1Y, function(x) sum(x>count.omics) > (prev.omics*length(x)), TRUE)
otu_table(fung.omics) <- microbiome::transform(otu_table(fung.omics), transform="clr")  
hist(as.vector(OTUDataframe(fung.omics)), breaks = 200, col = "#5C1E91")

# Save samples as column and features as rows
Bact.matrix <- OTUDataframe(bact.omics)
colnames(Bact.matrix) <- paste0("S", SampleDataframe(bact.1Y)$Subject)
Fung.matrix <- OTUDataframe(fung.omics)
colnames(Fung.matrix) <- paste0("S", SampleDataframe(fung.1Y)$Subject)
Vir.matrix <- OTUDataframe(vir.1Y)
colnames(Vir.matrix) <- paste0("S", SampleDataframe(vir.1Y)$Subject)

# Prepare nasal host matrix
Host.matrix.nasal <- host.nasal.1Y$E
colnames(Host.matrix.nasal) <- paste0('S', host.nasal.1Y$targets$Subjectnumber)
rownames(Host.matrix.nasal) <- make.unique(rownames(Host.matrix.nasal))
Host.matrix.nasal.immune <- Host.matrix.nasal[rownames(Host.matrix.nasal) %in% ImmuneDB$name,]

# Prepare blood host matrix
Host.matrix.blood <- host.dge.blood$E
colnames(Host.matrix.blood) <- paste0('S', host.dge.blood$targets$Subjectnumber)
rownames(Host.matrix.blood) <- make.unique(rownames(Host.matrix.blood))
Host.matrix.blood.immune <- Host.matrix.blood[rownames(Host.matrix.blood) %in% ImmuneDB$name,]

# Create list of omics datasets
omicslist <- list(Nasal=as.matrix(Host.matrix.nasal.immune), Blood=as.matrix(Host.matrix.blood.immune),
                  Bacteria=as.matrix(Bact.matrix), Fungi=as.matrix(Fung.matrix), Viruses=as.matrix(Vir.matrix))

# Create a MultiAssayExperiment object
omicsmae <- MultiAssayExperiment(experiments=omicslist, colData=DataFrame(metaS.subset))
omicsmae
```

#### 2) MOFA data integration

```{r}
# Create MOFA object
MOFAobject <- create_mofa(omicsmae)

# Set parameters
DataOptions <- get_default_data_options(MOFAobject)
ModelOptions <- get_default_model_options(MOFAobject)
ModelOptions$num_factors <- 3
TrainOptions <- get_default_training_options(MOFAobject)
TrainOptions$seed <- 2

# Run MOFA
MOFAobject_disease <- prepare_mofa(MOFAobject, data_options = DataOptions, model_options = ModelOptions, training_options = TrainOptions)
MOFArun <- run_mofa(MOFAobject_disease, use_basilisk = TRUE, outfile = "Output/MOFA_disease.rds")
```

#### 3) MOFA variance explained and factors

```{r}
# Number of top loadings to display
ntop <- 20

# Reload MOFA results
(plot.data.overview <- plot_data_overview(MOFAobject, colors = custom_palette_omics2, covariate = "Bacteria") +
   theme(legend.position = c(1, 0), legend.box.background=element_rect(), text=element_text(size=12)))
(plot.MOFA.var <- MOFAvarianceexplained(MOFArun, custom_palette_omics2, title = "CS-Group : MOFA variance explained") + theme(legend.position = "none"))
MOFArun@cache$variance_explained
MOFArun@cache$variance_explained$r2_per_factor

# Plot factor value with group
(plot.factor1.group <- BoxplotBinaryMOFAfactor(MOFArun, "Factor1", "Group", "Latent Factor1", custom_palette_2groups, custom_palette_2groups) + theme(legend.position = "none"))
(plot.factor1.daycare <- BoxplotBinaryMOFAfactor(MOFArun, "Factor1", "DayCare1YEAR", "Latent Factor1", custom_palette_2groups, c("grey", "grey")) + theme(legend.position = "none"))
 
# Stats for Factor1 values
factor1.values <- ReturnMOFAfactor2(MOFArun, "Factor1", "DayCare1YEAR", "Latent Factor1")
factor1.values <- factor1.values[factor1.values$Parameter!="None",]
factor1.values.wheeze <- factor1.values[factor1.values$Group == "Yes",]
factor1.values.ctrl <- factor1.values[factor1.values$Group == "No",]
wilcox.test(factor1.values$Factor ~ factor1.values$Parameter)
wilcox.test(factor1.values$Factor ~ factor1.values$Group)
wilcox.test(factor1.values.wheeze$Factor ~ factor1.values.wheeze$Parameter)
wilcox.test(factor1.values.ctrl$Factor ~ factor1.values.ctrl$Parameter)
factor1.values$Combined <- paste(factor1.values$Parameter, factor1.values$Group, sep="daycare_group")
pairwise.wilcox.test(factor1.values$Factor, factor1.values$Combined, p.adjust.method = "fdr")
kruskal.test(factor1.values$Factor ~ factor1.values$Combined)

# Plot
(plot.factor1.daycaregroup <- BoxplotBinaryMOFAfactor4groups(MOFArun, "Factor1", "DayCareGroup", "Latent Factor1", custom_palette_2groups, c(rep(custom_palette_2groups[1],2), rep(custom_palette_2groups[2],2)))+ theme(legend.position = "none"))

# Plot factor value with group
(plot.factor2.group <- BoxplotBinaryMOFAfactor(MOFArun, "Factor2", "Group", "Latent Factor2", custom_palette_2groups, custom_palette_2groups))
(plot.factor3.group <- BoxplotBinaryMOFAfactor(MOFArun, "Factor3", "Group", "Latent Factor3", custom_palette_2groups, custom_palette_2groups))
(plot.factor2.daycare <- BoxplotBinaryMOFAfactor(MOFArun, "Factor2", "DayCare1YEAR", "Latent Factor2", custom_palette_2groups, c("grey", "grey")))
(plot.factor3.daycare <- BoxplotBinaryMOFAfactor(MOFArun, "Factor3", "DayCare1YEAR", "Latent Factor3", custom_palette_2groups, c("grey", "grey")))

# Plot factor 1 top features
(plot.factor1.nasal.loadings <- MOFAfactorloadings(MOFArun, "Factor1", "Nasal", custom_palette_2groups, ntop, "CS-Group : Nasal host Factor1 MOFA loadings") + theme(legend.position = "none")) 
(plot.factor1.blood.loadings <- MOFAfactorloadings(MOFArun, "Factor1", "Blood", custom_palette_2groups[2], ntop, "CS-Group : Blood host Factor1 MOFA loadings") + theme(legend.position = "none"))
(plot.factor1.bact.loadings <- MOFAfactorloadings(MOFArun, "Factor1", "Bacteria", custom_palette_2groups, ntop, "CS-Group : Nasal bacteria Factor1 MOFA loadings") + theme(legend.position = "none"))
```

#### 4) Local-systemic top features commonalities

```{r}
# Get weights for both datasets
weights_nasal <- get_weights(MOFArun, views = "Nasal", factors = "Factor1", scale = FALSE)
weights_blood <- get_weights(MOFArun, views = "Blood", factors = "Factor1", scale = FALSE)

# Convert weights to data frames
weights_nasal_df <- as.data.frame(weights_nasal)
weights_blood_df <- as.data.frame(weights_blood)

# Add feature names as a column
weights_nasal_df$Feature <- rownames(weights_nasal_df)
weights_blood_df$Feature <- rownames(weights_blood_df)

# Remove suffixes from feature names
weights_nasal_df$Feature <- gsub("_Nasal|_Blood", "", weights_nasal_df$Feature)
weights_blood_df$Feature <- gsub("_Nasal|_Blood", "", weights_blood_df$Feature)

# Merge weights based on all feature names
merged_weights <- merge(weights_nasal_df, weights_blood_df, by = "Feature", suffixes = c("_Nasal", "_Blood"))

# Add absolute weights for both datasets
merged_weights$abs_weight_nasal <- abs(merged_weights$Factor1_Nasal)
merged_weights$abs_weight_blood <- abs(merged_weights$Factor1_Blood)

# Calculate combined rank based on absolute weights
merged_weights$combined_rank <- rank(-merged_weights$abs_weight_nasal) + rank(-merged_weights$abs_weight_blood)

# Select the top 8 features based on combined rank
top_8_features <- merged_weights %>%
  arrange(combined_rank) %>%
  head(8)

# Add a column to indicate if a feature is in the top 8
merged_weights$TopFeature <- ifelse(merged_weights$Feature %in% top_8_features$Feature, "Top 8", "Other")

# Define a custom color palette
palette <- c("Other" = "grey", "Top 8" = "lightsalmon2")

# Create scatter plot with correlation line
local.systemic.plot <- ggplot(merged_weights, aes(x = Factor1_Nasal, y = Factor1_Blood, label = Feature)) +
  geom_point(aes(fill = TopFeature, color = TopFeature), shape = 21, size = 2, stroke = 0.25) +
  geom_smooth(method = "lm", se = TRUE, color = "grey", fill = "grey", alpha = 0.2) +
  geom_label_repel(data = merged_weights[merged_weights$TopFeature == "Top 8", ], 
                   aes(label = Feature, fill = TopFeature), color = 'black', 
                   label.size = 0, label.r = 0.5, size = 4, 
                   label.padding = 0.25, max.overlaps = 30, 
                   segment.size = 0.25, segment.alpha = 0.95, segment.color = "grey") +
  labs(title = "Latent Factor1 correlations nasal-blood immune genes",
       x = "Nasal Weights",
       y = "Blood Weights") +
  scale_fill_manual(values = adjustcolor(palette, alpha = 0.5)) +
  scale_color_manual(values = adjustcolor(palette, alpha = 1)) +
  theme_minimal() +
  theme(legend.position = "none", legend.box.background = element_rect(), text = element_text(size = 12))

# Print the plot
print(local.systemic.plot)

# Save features
write.table(top_8_features$Feature, file = "Output/top_8_genes.txt", row.names = FALSE, quote = FALSE, col.names = FALSE)
```

#### 5) Monocytes markers

```{r}
# Define the common genes
common.nasal.blood <- top_8_features$Feature
# Define the markers
human.monocytes.markers <- c(
  "CD14", "CCR2", "CCL2", "CCL7",# Classical
  "FCGR3A", "CX3CR1", "CX3CL1")  # Non-Classical
# FCGR3A is CD16 (filtered out) 
# SELL is CD62L (filtered out)

# Extract gene expression matrices
nasal.common.genes <- as.matrix(t(Host.matrix.nasal[rownames(Host.matrix.nasal) %in% common.nasal.blood, ]))
nasal.monocytes.genes <- as.matrix(t(Host.matrix.nasal[rownames(Host.matrix.nasal) %in% human.monocytes.markers, ]))

# Perform correlation analyses
nasal.monocytes.cor <- psych::corr.test(nasal.common.genes, nasal.monocytes.genes, method = "spearman", adjust = "BH")

# Filter the markers present in the dataset
present_markers <- colnames(nasal.monocytes.genes)

# Define color mapping function
col_fun <- colorRamp2(c(-1, 0.25, 1), c("goldenrod2", "white", "#964067"))

# Define annotations for monocyte markers
marker_types <- c(
  "CD14" = "Classical", 
  "CCR2" = "Classical", 
  "CCL2" = "Classical", 
  "CCL7" = "Classical", 
  "CX3CR1" = "Non-Classical", 
  "FCGR3A" = "Non-Classical",
  "CX3CL1" = "Non-Classical")

marker_colors <- c("Classical" = "lightblue", "Intermediate" = "cadetblue3", "Non-Classical" = "steelblue")

# Filter the annotation data frame
annotation_df <- data.frame(
  Marker = present_markers,
  Type = marker_types[present_markers]
)

# Order the markers by their types
ordered_markers <- annotation_df$Marker[order(annotation_df$Type)]

# Create the heatmap for nasal monocytes correlations with space between different types
pdf(file = "Output/Figure6_Monocytes_markers.pdf", width = 6, height = 3 , onefile = T)
Heatmap(nasal.monocytes.cor$r[, ordered_markers, drop = FALSE], 
        name = "Correlation", 
        col = col_fun, 
        show_row_names = TRUE, 
        show_column_names = TRUE, 
        cluster_rows = TRUE, 
        cluster_columns = FALSE,
        column_split = annotation_df$Type[match(ordered_markers, annotation_df$Marker)],
        top_annotation = HeatmapAnnotation(
          Type = annotation_df$Type[match(ordered_markers, annotation_df$Marker)],
          col = list(Type = marker_colors)),
        column_names_rot = 0)
dev.off()
```

# [Single cell analysis]{.underline}

#### 1) HLCA pre-processing

```{r eval=FALSE}
# Interactive session
smux n -J Seurat --mem=300G --cpuspertask=1 --ntasks=1

# Load R
module load R/4.2.2-mkl

# Load required libraries
library(Seurat)
#library(SeuratDisk)
library(sctransform)
#library(glmGamPoi)
#library(sceasy)

# Load the HLCA core dataset
HLCA_core <- readRDS("local.rds")

# Check the available tissues in the metadata
table(HLCA_core@meta.data$tissue)

# Convert Ensembl IDs to gene symbols
Temp_Data <- GetAssayData(HLCA_core)
Temp_Meta <- HLCA_core@meta.data
new_rownames <- as.character(droplevels(HLCA_core@assays[["RNA"]]@meta.features[["feature_name"]]))
rownames(Temp_Data) <- new_rownames
HLCA_core <- CreateSeuratObject(counts = Temp_Data, meta.data = Temp_Meta)

# Remove cells with unknown annotation
HLCA_core <- subset(HLCA_core, subset = ann_finest_level != "Unknown")
HLCA_nose <- subset(HLCA_core, subset = tissue == "nose")

# Split the studies for batch correction
studies.list <- SplitObject(HLCA_nose, split.by = "study")

# Normalize and identify variable features for each study independently
studies.list <- lapply(X = studies.list, FUN = function(x) {
    x <- SCTransform(x, vst.flavor = "v2")
    x <- FindVariableFeatures(x, selection.method = "vst")
})

# Select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = studies.list, nfeatures = 2000)

# Run PCA on each dataset using these features
studies.list <- lapply(X = studies.list, FUN = function(x) {
    x <- ScaleData(x, features = features)
    x <- RunPCA(x, features = features)
})

# Find integration anchors
anchors <- FindIntegrationAnchors(object.list = studies.list, anchor.features = features, reduction = "rpca", k.anchor = 100)

# Integrate data
common.genes.df <- read.csv("top_8_genes.txt", header = FALSE)
features.interest <- c(features, common.genes.df[,1])
HLCA_nose_comb <- IntegrateData(anchorset = anchors, features.to.integrate = features.interest)
DefaultAssay(HLCA_nose_comb) <- "integrated"

# Run downstream analyses
HLCA_nose_comb <- ScaleData(HLCA_nose_comb)
HLCA_nose_comb <- RunPCA(HLCA_nose_comb, npcs = 20)
HLCA_nose_comb <- RunUMAP(HLCA_nose_comb, reduction = "pca", dims = 1:20)

# Subset myeloid cells
HCLA_nose_comb_myeloid <- subset(HLCA_nose_comb, subset = ann_level_2 == "Myeloid")

# Run downstream analyses
HCLA_nose_comb_myeloid <- RunPCA(HCLA_nose_comb_myeloid, npcs = 20)
HCLA_nose_comb_myeloid <- RunUMAP(HCLA_nose_comb_myeloid, reduction = "pca", dims = 1:20)

# Save rds files
saveRDS(HCLA_nose_comb_myeloid, "HCLA_nose_comb_myeloid_core_export.rds")
saveRDS(HLCA_nose_comb, "HLCA_nose_comb_core_export.rds")
```

#### 2) UMAP and dotplots of genes of interest

```{r}
# Load datasets
HCLA_data_nose <- readRDS("Output/HLCA_nose_comb_core_export.rds")
HCLA_data_nose_myeloid <- readRDS("Output/HCLA_nose_comb_myeloid_core_export.rds")
```

```{r}
common.nasal.blood <- top_8_features$Feature

# Single cell palette
custom_palette_singlecell1 <- c(paletteer_dynamic("cartography::sand.pal", 3), paletteer_dynamic("cartography::wine.pal", 3), "lightpink3", "grey")
custom_palette_singlecell2 <- c(paletteer_dynamic("cartography::sand.pal", 2), paletteer_dynamic("cartography::wine.pal", 3))

# Harmonize airway epithelium label
HCLA_data_nose@meta.data$ann_level_2.2 <- ifelse(HCLA_data_nose@meta.data$ann_level_2 == "Alveolar epithelium", "Airway epithelium", HCLA_data_nose@meta.data$ann_level_2)

# Select major cell types (>1000 cells)
HCLA_data_nose <- subset(HCLA_data_nose, subset = ann_level_2 == names(which(table(HCLA_data_nose@meta.data$ann_level_2) > 1000)))

# Plot all nasal cells
(allcells.ann2.umap <- DimPlot(HCLA_data_nose, group.by = "ann_level_2.2", label = TRUE, repel = TRUE, label.size = 5, raster = FALSE,
                               cols = custom_palette_singlecell1) +
    labs(title = "HCLA core : nasal tissue cells") +
    theme(axis.text.x=element_blank(),
          legend.position = "none",
          axis.text.y=element_blank(), axis.ticks=element_blank(),
          legend.box.background=element_rect(), text=element_text(size=12)))

# Check expression of our genes of interest
(allcells.ann2.dotplot <- DotPlot(HCLA_data_nose, features = common.nasal.blood, dot.min = 0, group.by = "ann_level_2.2", dot.scale = 10, scale.max = 50, scale.by = "size") +
    labs(title = "Local-systemic Factor1 gene set expression in nasal cells subset") +
    geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0) +
    scale_color_gradient2(low = "grey80", mid = "white", high = "#964091") +
    theme(legend.box.background=element_rect(), axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none",
          text=element_text(size=12)))

# Subset myeloid cells
HCLA_data_nose_myeloid <- FindNeighbors(HCLA_data_nose_myeloid, dims = 1:10)
HCLA_data_nose_myeloid <- FindClusters(HCLA_data_nose_myeloid , resolution = 0.4, graph.name = "integrated_nn", initial.membership = )

# Reannotate clusters
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- NA
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==0, "DC2", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==1, "Macrophages", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==2, "DC1", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==3, "Plasmacytoid DCs", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==4, "Mast cells", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==5, "Monocytes", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)
HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual <- ifelse(HCLA_data_nose_myeloid@meta.data$seurat_clusters==6, "Migratory DCs ", HCLA_data_nose_myeloid@meta.data$seurat_clusters_manual)

# Plot myeloid cells
(myeloid.anno.umap <- DimPlot(HCLA_data_nose_myeloid, group.by = "seurat_clusters_manual", label = TRUE, repel = TRUE, label.size = 5,
                             reduction = "umap", cols = custom_palette_singlecell1) +
    labs(title = "HCLA core : nasal myeloid cells") +
    theme(axis.text.x=element_blank(), axis.text.y=element_blank(), axis.ticks=element_blank(),
          legend.position = "none",
          legend.box.background=element_rect(), text=element_text(size=12)))

# Create dotplots for common genes
(myeloid.anno.dotplot <- DotPlot(HCLA_data_nose_myeloid, features = common.nasal.blood, dot.min = 0, group.by = "seurat_clusters_manual", dot.scale = 10, scale.max = 50,
                                 scale.by = "size", cols = scales::alpha(c("grey80", "#964091"), 0.75)) +
    labs(title = "Local-systemic Factor1 gene set expression in nasal myeloid cells subset") +
    geom_point(aes(size=pct.exp), shape = 21, colour="black", stroke=0) +
    scale_color_gradient2(low = "grey80", mid = "white", high = "#964091") +
    theme(legend.box.background=element_rect(), axis.text.x = element_text(angle = 90, hjust = 1),
          legend.position = "none",
          text=element_text(size=12)))
```

```{r}
# Step 1: Create a combined annotation column
HCLA_data_nose_myeloid@meta.data$combined_annotation <- ifelse(
  HCLA_data_nose_myeloid@meta.data$ann_level_4 == "None",
  HCLA_data_nose_myeloid@meta.data$cell_type,
  HCLA_data_nose_myeloid@meta.data$ann_level_4
)

# Step 2: Create the table using the combined annotation
annotation_table <- table(HCLA_data_nose_myeloid@meta.data$combined_annotation, 
                          HCLA_data_nose_myeloid@meta.data$seurat_clusters)

# Convert table to matrix for easier manipulation
annotation_matrix <- as.matrix(annotation_table)

# Step 3: Function to assign unique labels to clusters
assign_unique_labels <- function(data) {
  n_clusters <- ncol(data)
  labels <- character(n_clusters)
  used_annotations <- character(0)
  
  for (i in 1:n_clusters) {
    sorted_indices <- order(data[, i], decreasing = TRUE)
    
    for (j in sorted_indices) {
      candidate_label <- rownames(data)[j]
      if (candidate_label %in% used_annotations || candidate_label == "None") {
        next
      } else {
        labels[i] <- candidate_label
        used_annotations <- c(used_annotations, candidate_label)
        break
      }
    }
    
    if (labels[i] == "") {
      labels[i] <- "Unassigned"
    }
  }
  
  return(labels)
}

# Step 4: Apply the function to our matrix
cluster_labels <- assign_unique_labels(annotation_matrix)

# Step 5: Print the results
for (i in 1:length(cluster_labels)) {
  cat("Cluster", colnames(annotation_matrix)[i], ": ", cluster_labels[i], "\n")
}
```

### ***Figure 2***

Longitudinal analysis of all omics.

```{r}
# Set size
label.size <- 20

(plot1 <- ggarrange(plot.nasal.PCA.Age, NA, bact.ordi.long, NA, fung.ordi.long,
                    ncol = 5, nrow = 1, align = "hv",
                    widths = c(1, 0.2, 1, 0.2, 1),
                    labels = c("a","","b", "", "c"),
                    font.label = list(size=label.size)))

(plot2 <-  ggarrange(host.long.distance, NA, bact.long.distance, NA, fung.long.distance,
                    ncol = 5, nrow = 1, align = "hv",
                    widths = c(1, 0.2, 1, 0.2, 1),
                    labels = c("d","", "e", "", "f"),
                    font.label = list(size=label.size)))

(plot3 <- ggarrange(bact.Observed.age, NA, fung.Observed.age, NA, virus.plot.detection.long, NA, virus.plot.details.long, 
                    ncol = 7, nrow = 1, align = "hv",
                    widths = c(1.25, 0.2, 1.25, 0.2, 1, 0.2, 1),
                    labels = c("g","", "h", "", "i", "", "j"),
                    font.label = list(size=label.size)))

(plot4 <- ggarrange(plot.volcano.Age, NA, plot.immune.pathway.age, NA, pathways.long.age, 
                    ncol = 5, nrow = 1, align = "hv",
                    widths = c(2, 0.2, 2, 0.2, 1.5),
                    labels = c("k","", "l", "", ""),
                    font.label = list(size=label.size)))

plot5 <- ggarrange(pathways.plot.jakstatdown, pathways.plot.jakstatupn, pathways.plot.Bcell,
                   pathways.plot.Th1Th2, pathways.plot.th17,
                    ncol = 6, nrow = 1, align = "hv",
                    widths = c(1, 1, 1, 1, 1, 1),
                    labels = c("m", "n", "o", "p", "q"),
                    font.label = list(size=label.size))

plot6 <- ggarrange(bact.long.age.bar, NA, 
                    ncol = 2, nrow = 1, align = "hv",
                    widths = c(1, 2),
                    labels = c("r", ""),
                    font.label = list(size=label.size))

Figure2 <- ggarrange(plot1, NULL, plot2, NULL, plot3, NULL, plot4, NULL, plot5, NULL, plot6,
                      ncol = 1, nrow = 11, align = "hv",
                      heights = c(1, 0.1, 1, 0.1, 1, 0.1, 1.2, 0.1, 1.2, 0.1, 1.2),
                      font.label = list(size=label.size))

# Save figure
ggsave("Output/Figure2.pdf", Figure2, width=45, height=60, units="cm", useDingbats=FALSE)
```

### ***Figure 3***

Covariates influencing the different omics at age 1.

```{r}
# Set size
label.size <- 20

# Create plots
(plot1 <- all.cor.factors) 

(plot2 <- ggarrange(alpha.bact.daycare, NA, alpha.vir.daycare,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.25, 1),
                    labels = c("b","","c"),
                    font.label = list(size=label.size)))

(plot3 <- ggarrange(alpha.bact.abx, NA, alpha.vir.abx,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.25, 1),
                    labels = c("d", "", "e"),
                    font.label = list(size=label.size)))

(plot23 <- ggarrange(plot2, NA, plot3,
                     ncol = 1, nrow = 3, align = "hv",
                     heights = c(1, 0.1, 1),
                     font.label = list(size=label.size)))

(plot123 <- ggarrange(plot1, NA, plot23,
                     ncol = 3, nrow = 1, align = "hv",
                     labels = c("a", "", ""),
                     widths = c(1, 0.25, 1.5),
                     font.label = list(size=label.size)))

(plot4 <- ggarrange(host.nasal.covariates.plot, NA, host.blood.covariates.plot, 
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.25, 1),
                    labels = c("f","","g"),
                    font.label = list(size=label.size)))

(plot5 <- ggarrange(covariates.da.bact.plot, NA, bact.da.daycare.bar,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.25, 1),
                    labels = c("h","","i"),
                    font.label = list(size=label.size)))

(plot6 <- ggarrange(covariates.da.fung.plot, NA,  fung.da.season.bar,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.25, 1),
                    labels = c("j","","k"),
                    font.label = list(size=label.size)))

(Figure3 <- ggarrange(plot123, NA, plot4, NA, plot5, NA, plot6,
                      ncol = 1, nrow = 7, align = "hv",
                      heights = c(1.5, 0.1, 1, 0.1, 1, 0.1, 1),
                      font.label = list(size=label.size)))

# Save figure
ggsave("Output/Figure3.pdf", Figure3, width=25, height=55, units="cm", useDingbats=FALSE)
```

### ***Figure 4***

Wheeze 1 year cohort.

```{r}
# Set size
label.size <- 20

# Create plots
(plot1 <- ggarrange(bact.Observed.1Y, NA, fung.Observed.1Y, 
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.1, 1),
                    labels = c("a", "", "b"),
                    font.label = list(size=label.size)))

(plot2 <- ggarrange(bact.da.disease, NA, bact.da.disease.adj, 
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.1, 1),
                    labels = c("c", "", "d"),
                    font.label = list(size=label.size)))

(plot3 <- ggarrange(virus.plot.wheeze, NA, virus.plot.details.wheeze,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(1, 0.1, 1),
                    labels = c("f", "", "g"),
                    font.label = list(size=label.size)))
  
(plot4 <- plot.volcano.nasal.Disease)

Figure4 <- ggarrange(plot1, NA, plot2, NA, NA, plot3, NA, plot4,
                     ncol = 1, nrow = 8, align = "hv",
                     heights = c(1, 0.1, 1, 0.1, 1, 1, 0.1, 1),
                     labels = c("", "", "", "", "e", "", "", "", "h"),
                     font.label = list(size=label.size))
# Save figure
ggsave("Output/Figure4.pdf", Figure4, width=22, height=52, units="cm", useDingbats=FALSE)
```

### ***Figure 5***

```{r}
# Set size
label.size <- 20

# Create plots
(plot1 <- ggarrange(plot.MOFA.var, NA, plot.factor1.group, NA, plot.factor1.daycaregroup,
                    ncol = 5, nrow = 1, align = "hv",
                    widths = c(2, 0, 1.5, 0, 2),
                    labels = c("a","","b", "", "c"),
                    font.label = list(size=label.size)))

(plot2 <- ggarrange(plot.factor1.nasal.loadings, NA, plot.factor1.blood.loadings, NA, plot.factor1.bact.loadings,
                    ncol = 5, nrow = 1, align = "hv",
                    widths = c(1, 0, 1, 0, 1.5),
                    labels = c("d","","e", "", "f"),
                    font.label = list(size=label.size)))

plot3 <- ggarrange(allcells.ann2.umap, NA, allcells.ann2.dotplot,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(2.5, 0.1, 3),
                    labels = c("g","","h"),
                    font.label = list(size=label.size))

plot4 <- ggarrange(myeloid.anno.umap, NA, myeloid.anno.dotplot,
                    ncol = 3, nrow = 1, align = "hv",
                    widths = c(2.5, 0.1, 3),
                    labels = c("i","","j"),
                    font.label = list(size=label.size))

Figure5 <- ggarrange(plot1, NA, plot2, NA, plot3, NA, plot4, NA, NA,
                     ncol = 1, nrow = 9, align = "hv",
                     heights = c(2, 0.1, 2.5, 0.1, 2.2, 0.1, 2.2, 0.1, 2),
                     labels = c("","","","","","","","","k"),
                     font.label = list(size=label.size))
# Save figure
ggsave("Output/Figure5.pdf", Figure5, width=36, height=52, units="cm", useDingbats=FALSE)
```


